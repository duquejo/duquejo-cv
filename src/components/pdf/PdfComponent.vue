
<template>
  <button @click.stop="handleGeneratePdf"
    class="group animate-bounce-subtle inline-flex items-center justify-center text-xs overflow-hidden rounded-2xl px-6 font-medium hover:bg-gray-100 hover:text-black bg-gray-100 leading-none py-0">
    <div class="relative inline-flex -translate-x-0 items-center transition group-hover:-translate-x-6">
      <div class="absolute translate-x-0 opacity-100 transition group-hover:-translate-x-6 group-hover:opacity-0">
        <v-icon name="md-pictureaspdf" scale="0.9" />
      </div>
      <span class="pl-6 w-24">{{ t('general.services.download_button') }}</span>
      <div class="absolute right-0 translate-x-12 opacity-0 transition group-hover:translate-x-6 group-hover:opacity-100">
        <v-icon name="ri-download-cloud-2-line" scale="0.9" />
      </div>
    </div>
  </button>
</template>

<script setup lang="ts">
import { generate } from '@pdfme/generator';
import { Template } from '@pdfme/common';
import { text } from '@pdfme/schemas';

import { BASE_PDF_ES, BASE_PDF_EN } from '@/shared/constants';
import { schema } from '@/components/pdf/config';
import useLanguageContext from '@/composables/useLanguageContext';
import { calculateYears } from '@/shared/helpers/calculateYears';

const { t, languageSources, locale } = useLanguageContext();

const handleGeneratePdf = () => {

  const template: Template = {
    basePdf: locale.value === 'es' ? BASE_PDF_ES : BASE_PDF_EN,
    schemas: schema,
  };

  const plugins = { text };

  const inputs = [{
    introduction: languageSources.value.pdf.introduction,
    experience: `${calculateYears.value} ${languageSources.value.pdf.experience}`,
    languages: languageSources.value.languages.map(lang => `${lang.title} - ${lang.content}`).join('\n'),
    databases: languageSources.value.pdf.databases,
    tools: languageSources.value.pdf.cloud,
    backend: languageSources.value.pdf.backend,
    cicd: languageSources.value.pdf.cicd,
    frontend: languageSources.value.pdf.frontend,
    architecture: languageSources.value.pdf.architecture,
    programming_languages: languageSources.value.pdf.programming_languages,
    methodologies: languageSources.value.pdf.methodologies,
    security: languageSources.value.pdf.security,
    email_bottom: 'duquejo01@gmail.com',
    website_bottom: document.location.origin,
  }];

  generate({ template, plugins, inputs }).then((pdf) => {
    // Browser
    const blob = new Blob([pdf.buffer], { type: 'application/pdf' });
    window.open(URL.createObjectURL(blob));
  });
}
</script>