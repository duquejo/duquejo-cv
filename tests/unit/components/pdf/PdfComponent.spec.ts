import { beforeEach, describe, expect, it, vi } from 'vitest';
import { VueWrapper, shallowMount } from '@vue/test-utils';

import { useI18n } from 'vue-i18n';

import PdfComponent from '../../../../src/components/pdf/PdfComponent.vue';
import createVueI18n from '../../../mocks/mock-i18n';

vi.mock('vue-i18n', async () => {
  const actual = await vi.importActual('vue-i18n');
  return {
    ...actual,
    useI18n: vi.fn(),
  };
});

const mockUseI18n = vi.mocked(useI18n);

describe('+ PdfComponent.vue tests', () => {

  const i18n = createVueI18n();

  let wrapper: VueWrapper;

  beforeEach(() => {
    vi.clearAllMocks();
    mockUseI18n.mockReturnValue({
      t: (key: string) => key,
      locale: vi.fn(),
      getLocaleMessage: (key: string) => vi.fn(),
    });
    wrapper = shallowMount(PdfComponent, {
      global: {
        plugins: [i18n],
      }
    });
  });

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(mockUseI18n).toHaveBeenCalled();
    expect(mockUseI18n).toHaveBeenCalledWith();
  });

  it('should have visible the PDF button with the ready state', () => {
    expect(wrapper.find('button').attributes('disabled')).toBeFalsy();
    expect(wrapper.find('.cursor-wait').exists()).toBeFalsy();
  });

  it('Should validate the PDF generation inner statuses', async () => {
    const btn = wrapper.find('button');
    await btn.trigger('click');
    expect(wrapper.find('.cursor-wait').exists()).toBeTruthy();
    expect(btn.attributes('disabled')).toBe('');
  });
});