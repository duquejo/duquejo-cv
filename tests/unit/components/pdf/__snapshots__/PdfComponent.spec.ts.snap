// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ PdfComponent.vue tests > should match the snapshot 1`] = `
"<button class="group animate-bounce-subtle inline-flex items-center justify-center text-xs overflow-hidden rounded-2xl px-6 font-medium hover:bg-gray-100 hover:text-black bg-gray-100 leading-none py-0">
  <div class="relative inline-flex -translate-x-0 items-center transition group-hover:-translate-x-6">
    <div class="absolute translate-x-0 opacity-100 transition group-hover:-translate-x-6 group-hover:opacity-0">
      <v-icon name="md-pictureaspdf" scale="0.9"></v-icon>
    </div><span class="pl-6 w-24">general.services.download_button</span>
    <div class="absolute right-0 translate-x-12 opacity-0 transition group-hover:translate-x-6 group-hover:opacity-100">
      <v-icon name="ri-download-cloud-2-line" scale="0.9"></v-icon>
    </div>
  </div>
</button>"
`;
